database/sqlite/sqlite_aclk.c

```diff
@@ -594,206 +408,216 @@ static void aclk_database_worker(void *arg)
             host->node_id = mallocz(sizeof(*host->node_id));
             uuid_copy(*host->node_id, *node_id);
         }
     }
-    else
-        wc->hostname = get_hostname_by_node_id(wc->node_id);
     wc->host = host;
     strcpy(wc->uuid_str, uuid_str);
-    strcpy(wc->host_guid, host_guid);
     wc->alert_updates = 0;
-    aclk_database_init_cmd_queue(wc);
-    aclk_add_worker_thread(wc);
-    fatal_assert(0 == uv_thread_create(&(wc->thread), aclk_database_worker, wc));
-#else
-    UNUSED(host);
-    UNUSED(host_uuid);
-    UNUSED(node_id);
+    wc->node_info_send = 1;
 #endif
-}
\ No newline at end of file
+}
+
+void sql_aclk_sync_init(void)
+{
+    char *err_msg = NULL;
+    int rc;
+
+    if (unlikely(!db_meta)) {
+        if (default_rrd_memory_mode != RRD_MEMORY_MODE_DBENGINE) {
+            return;
+        }
+        error_report("Database has not been initialized");
+        return;
+    }
+
+    info("Creating archived hosts");
+    rc = sqlite3_exec_monitored(db_meta, "SELECT host_id, hostname, registry_hostname, update_every, os, "
+                                         "timezone, tags, hops, memory_mode, abbrev_timezone, utc_offset, program_name, "
+                                         "program_version, entries, health_enabled FROM host WHERE hops >0;",
+                                create_host_callback, NULL, &err_msg);
+
+    if (rc != SQLITE_OK) {
+        error_report("SQLite error when loading archived hosts, rc = %d (%s)", rc, err_msg);
+        sqlite3_free(err_msg);
+    }
+    // Trigger host context load for hosts that have been created
+    metadata_queue_load_host_context(NULL);
+
+#ifdef ENABLE_ACLK
+    rc = sqlite3_exec_monitored(db_meta, "SELECT ni.host_id, ni.node_id FROM host h, node_instance ni " \
+        "WHERE h.host_id = ni.host_id AND ni.node_id IS NOT NULL;",
+        aclk_config_parameters, NULL,&err_msg);
+
+    if (rc != SQLITE_OK) {
+        error_report("SQLite error when configuring host ACLK synchonization parameters, rc = %d (%s)", rc, err_msg);
+        sqlite3_free(err_msg);
+    }
+    aclk_synchronization_init();
+
+    info("ACLK sync initialization completed");
+#endif
+}
+
+// Public
+
+static inline void queue_aclk_sync_cmd(enum aclk_database_opcode opcode, const void *param0, const void *param1)
+{
+    struct aclk_database_cmd cmd;
+    cmd.opcode = opcode;
+    cmd.param[0] = (void *) param0;
+    cmd.param[1] = (void *) param1;
+    cmd.completion = NULL;
+    aclk_database_enq_cmd(&cmd);
+}
+
+// Public
+void aclk_push_alert_config(const char *node_id, const char *config_hash)
+{
+    if (unlikely(!aclk_sync_config.initialized))
+        return;
+
+    queue_aclk_sync_cmd(ACLK_DATABASE_PUSH_ALERT_CONFIG, node_id, config_hash);
+}
+
+void aclk_push_node_alert_snapshot(const char *node_id)
+{
+    if (unlikely(!aclk_sync_config.initialized))
+        return;
+
+    queue_aclk_sync_cmd(ACLK_DATABASE_PUSH_ALERT_SNAPSHOT, strdupz(node_id), NULL);
```

Comment:
Are we not having memory leaks here? I remember to chat with @MrZammler about this one day, but I am not sure if it was with the same function.

Reply:
If this is happening with this function, it is also happening in the nexts.

Reply:
> 

Good catch.. there is a memory leak under certain condition. I will fix